### PR TITLE
[RFC] Add CG test using cuda

### PR DESCRIPTION
**Pull request by Rombur**
This PR adds a test where `dealii::SolverCG` is used with a CUDA matrix. However there is a problem with `GrowingVectorMemory`. Basically what happens is that `GrowingVectorMemory` release the memory after the CUDA environment is shut down and so we get an error about the driver shutting down (see [here](https://stackoverflow.com/questions/43235548/how-to-handle-cudafree-on-globally-instantiated-variables) for a similar problem). The solution is to free the memory by hand before exiting the program. So I have two questions:
 1. Is this an acceptable solution and if not, what should I do?
 2. Giving this problem should we instantiation `GrowingVectorMemory` for the CUDA vector or this should be done by more advanced users only?

### CODE REVIEW

**Comment by bangerth** [comment]
Do we need an object like `MPI_InitFinalize` that handles the set-up/tear-down of CUDA in the same way as we do MPI? If so, we could call the `.release()` function for the CUDA vectors in the destructor of that object?

**Comment by bangerth** [comment]
If that's too complicated, I'm definitely ok with the explicit variant here as well.

**Comment by tjhei** [comment]
``~MPI_InitFinalize`` already deletes memory from the ``GrowingVectorMemory<T>`` for all the parallel vectors. Wouldn't it make sense to just add CUDA here?

**Comment by Rombur** [comment]
All the CUDA libraries need to create and destroy a handle that you need to pass around. Right now  we only use one handle but I will need more in the future. So what about I do something like this:
```c++
struct CUDAHandle
{
  CUDAHandle() {/* create the handles */}
  ~CUDAHandle() {/* free the memory and destroy the handles */}
  cudaHandle_1 handle_1;
  cudaHandle_2 handle_2;
  cudaHandle_3 handle_3;
};
```
Then we would pass this object around instead of the individual handle and when the object is destroyed I can free the memory. The only weird thing is that if you are doing something matrix-free, you won't need the handle but you still need to create the object to free the memory correctly.

**Comment by bangerth** [comment]
On 04/11/2018 07:13 AM, Timo Heister wrote:
> |~MPI_InitFinalize| already deletes memory from the |GrowingVectorMemory<T>| 
> for all the parallel vectors. Wouldn't it make sense to just add CUDA here?

That would make sense -- once you shut down the MPI system, one can presume 
that (i) you don't want to do any more computations, (ii) none of the vectors 
are still in use. So releasing the memory in |~MPI_InitFinalize| seems sensible.

At the same time, one can do CUDA computations without doing MPI. So there 
needs to be an alternative way for those cases as well.

In other words, we should probably do both.

**Comment by bangerth** [comment]
@Rombur -- about your example, is there a CUDA thing that says that you only need 3 handles? I have no background in CUDA programming, so can't say whether that makes sense, of course...

**Comment by bangerth** [comment]
Independent of this conversation, though, should we just merge the current testcase? It is possible to come up with a better solution, but there is nothing wrong with the testcase itself.

**Comment by Rombur** [comment]
> @Rombur -- about your example, is there a CUDA thing that says that you only need 3 handles? I have no background in CUDA programming, so can't say whether that makes sense, of course...

Basically you need a handle for each cuda library that you are using. So the number of handles will depend on the number of libraries we will use. Right now we use only cuSPARSE in deal but in my code I am also using cuSOLVER dense and cuSOLVER sparse. I plan to use at least cuSOLVER sparse in deal but there are more libraries that are using handles like cuBLAS and cuFFT. So I don't know how many will use in the future but at least two.

> Independent of this conversation, though, should we just merge the current testcase? It is possible to come up with a better solution, but there is nothing wrong with the testcase itself.

That's fine, we can merge this and I'll change the test later.

**Comment by bangerth** [comment]
I see. That makes destroying these handles more complicated.

I think that @tjhei 's suggesting of putting this into `~MPI_InitFinalize` is a good idea. It at least covers the case where someone is using CUDA *and* MPI.

**Comment by Rombur** [comment]
The problem is that creating the handle is not enough. Every time we need to call a function from a library we need to pass the handle to the function. Let's say I want to do a `vmult`, I need to do something like this `vmult(cusparse_handle, matrix, src, dst)`. So I need to be able to pass the handle around or it needs to be a global variable.

**Comment by Rombur** [comment]
> How do other codes handle these things?

I have no idea :-(

**Comment by Rombur** [comment]
> When would you need more than one handle? If you don't it could just be a "global variable" (or returned by a static function) somewhere.

So far I use only one handle per library and I don't have a case where we would want to use multiple handles of the same library. I know it is recommended to have multiple handles for multi-threading but it's to prevent user error not because of limitation in the libraries. I am not a fan of global variable but that would simplify the interface of the functions. What do you guys prefer?

**Comment by tjhei** [comment]
> What do you guys prefer?

If we don't think more than 1 handle is needed, I would prefer not having to pass it around.

**Comment by Rombur** [comment]
I looked at what they do in `TensorFlow` and they create a structure similar to the one I proposed. They also have multiple handles for each library because they use multiple streams. Streams allow you to overlap communication and data transfer. Using multiple streams is somewhat similar to use non-blocking functions in MPI.
